File: eval/starters/unity/Assets/Editor/Probe.cs
// Batchmode entry points for the probe protocol. See `## Probing a run` in
// AGENTS.md for the contract these implement.
//
// Three of them, all driven by `-executeMethod`:
//
//   Probe.Stream    `just probe`       stdin/stdout, one tick per input line
//   Probe.Batch     `just probe-file`  a whole scripted run to a JSONL file
//   Probe.Film      `just film`        evenly spaced PNGs of the same run
//
// Traps this file exists to avoid:
//
// * A batchmode run that throws still exits 0 on some paths, so every entry
//   point wraps its body and calls `EditorApplication.Exit` with an explicit
//   code. Silence is not success.
// * `Stream` must NOT be launched with `-quit`: the editor would exit before a
//   single line was read. `Batch` and `Film` do use `-quit`, which is safe —
//   the ban is specifically `-quit` together with `-runTests`.
// * stdout carries the trace and nothing else. Diagnostics go to stderr, and
//   the engine's own chatter goes to the `-logFile`.

using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Text;
using Starter.Sim;
using Starter.View;
using UnityEditor;

namespace Starter.EditorTools
{
    public static class Probe
    {
        /// How many frames `Film` writes at most, including both endpoints.
        private const int FilmFrames = 12;

        // ------------------------------------------------------------------
        // Entry points
        // ------------------------------------------------------------------

        /// Long-lived headless process: one JSON input object per stdin line in,
        /// one JSON trace line per tick out, flushed every time.
        public static void Stream()
        {
            int code = 0;
            try
            {
                var state = new SimState(ULongArg("-seed"));
                using var stdout = OpenTrace();

                // The world before any tick has run, so a driver can decide what
                // to do with the very first input.
                Emit(stdout, state);

                string line;
                while ((line = Console.In.ReadLine()) != null)
                {
                    if (string.Equals(line.Trim(), "quit", StringComparison.Ordinal)) break;
                    state.Step(IntentsFrom(line));
                    Emit(stdout, state);
                }
            }
            catch (Exception e)
            {
                Console.Error.WriteLine("probe: " + e);
                code = 1;
            }
            EditorApplication.Exit(code);
        }

        /// Run a scripted number of ticks and write the whole trace to a file.
        public static void Batch()
        {
            int code = 0;
            try
            {
                ulong seed = ULongArg("-seed");
                int ticks = IntArg("-ticks");
                var inputs = ReadScript(StringArg("-script"));
                string outPath = StringArg("-out");

                var state = new SimState(seed);
                var text = new StringBuilder();
                for (int tick = 0; tick < ticks; tick++)
                {
                    state.Step(tick < inputs.Count ? inputs[tick] : Intents.None);
                    text.Append(TraceLine(state)).Append('\n');
                }

                if (state.Tick != (ulong)ticks)
                {
                    throw new InvalidOperationException(
                        $"asked for {ticks} ticks, the simulation reported {state.Tick}");
                }

                string dir = Path.GetDirectoryName(Path.GetFullPath(outPath));
                if (!string.IsNullOrEmpty(dir)) Directory.CreateDirectory(dir);
                File.WriteAllText(outPath, text.ToString());
                Console.Error.WriteLine(
                    $"probe-file: wrote {ticks} lines to {outPath}");
            }
            catch (Exception e)
            {
                Console.Error.WriteLine("probe-file: " + e);
                code = 1;
            }
            EditorApplication.Exit(code);
        }

        /// Capture up to twelve evenly spaced frames of the same scripted run.
        public static void Film()
        {
            int code = 0;
            try
            {
                ulong seed = ULongArg("-seed");
                int ticks = IntArg("-ticks");
                var inputs = ReadScript(StringArg("-script"));
                string outDir = Path.GetFullPath(StringArg("-outdir"));
                Directory.CreateDirectory(outDir);

                var script = inputs.ToArray();
                int written = 0;
                int previous = -1;
                foreach (int tick in FrameTicks(ticks))
                {
                    if (tick == previous) continue;
                    previous = tick;
                    var frame = RenderHarness.CaptureFrame(seed, tick, script);
                    frame.SavePng(Path.Combine(
                        outDir,
                        "frame_" + written.ToString("0000", CultureInfo.InvariantCulture) + ".png"));
                    written++;
                }
                Console.Error.WriteLine($"film: wrote {written} frame(s) to {outDir}");
            }
            catch (Exception e)
            {
                Console.Error.WriteLine("film: " + e);
                code = 1;
            }
            EditorApplication.Exit(code);
        }

        /// Where the trace goes.
        ///
        /// NOT `Console.Out`, and not file descriptor 1 either: `-logFile`
        /// makes batchmode Unity redirect the process's own stdout into the log,
        /// so anything written there would be buried in the engine banner and
        /// the driver would see an empty pipe. `just probe` therefore hands the
        /// real stdout through as descriptor 3 (`3>&1`) and passes its path in
        /// `-trace`. Without that argument the trace goes to descriptor 1, which
        /// is right when nothing has redirected it.
        private static StreamWriter OpenTrace()
        {
            string path = OptionalArg("-trace");
            var stream = string.IsNullOrEmpty(path)
                ? Console.OpenStandardOutput()
                : new FileStream(path, FileMode.Open, FileAccess.Write);
            return new StreamWriter(stream, new UTF8Encoding(false)) { AutoFlush = false };
        }

        /// The ticks `Film` captures: `FilmFrames` samples spread over
        /// `0..=ticks` inclusive, endpoints always included, integer division so
        /// the spacing is reproducible rather than float-rounded.
        internal static IEnumerable<int> FrameTicks(int ticks)
        {
            if (ticks <= 0) { yield return 0; yield break; }
            for (int i = 0; i < FilmFrames; i++)
            {
                yield return (int)((long)i * ticks / (FilmFrames - 1));
            }
        }

        // ------------------------------------------------------------------
        // The trace format
        // ------------------------------------------------------------------

        private static void Emit(StreamWriter stdout, SimState state)
        {
            stdout.Write(TraceLine(state));
            stdout.Write('\n');
            // Mandatory. A block-buffered stdout deadlocks a driver that is
            // waiting for this line before it sends the next input.
            stdout.Flush();
        }

        /// One line of the trace. `state` is game-defined; this is the
        /// placeholder's version of it.
        private static string TraceLine(SimState state)
        {
            var marker = state.Marker;
            var sb = new StringBuilder(160);
            sb.Append("{\"tick\": ").Append(state.Tick.ToString(CultureInfo.InvariantCulture));
            sb.Append(", \"hash\": \"0x")
              .Append(state.StateHash().ToString("x16", CultureInfo.InvariantCulture))
              .Append('"');
            sb.Append(", \"state\": {\"marker\": {\"x\": ").Append(Number(marker.Position.X))
              .Append(", \"y\": ").Append(Number(marker.Position.Y))
              .Append(", \"vx\": ").Append(Number(marker.Velocity.X))
              .Append(", \"vy\": ").Append(Number(marker.Velocity.Y))
              .Append("}}");

            sb.Append(", \"events\": [");
            var names = state.Events.Names;
            for (int i = 0; i < names.Count; i++)
            {
                if (i > 0) sb.Append(", ");
                sb.Append('"').Append(names[i]).Append('"');
            }
            sb.Append("]}");
            return sb.ToString();
        }

        /// Nine significant digits round-trips a float exactly. NaN and infinity
        /// are not JSON numbers, so they fail the run instead of producing a
        /// trace no parser can read.
        private static string Number(float value)
        {
            if (float.IsNaN(value) || float.IsInfinity(value))
            {
                throw new InvalidOperationException(
                    $"the simulation produced a non-finite value ({value}); a trace cannot " +
                    "represent it. Find the division by zero.");
            }
            return value.ToString("G9", CultureInfo.InvariantCulture);
        }

        // ------------------------------------------------------------------
        // Input
        // ------------------------------------------------------------------

        /// One stdin line to one tick of intent. An empty line means "nothing
        /// held", which is what a driver that has nothing to say should send.
        private static Intents IntentsFrom(string line)
        {
            if (string.IsNullOrWhiteSpace(line)) return Intents.None;
            return IntentsFrom(Json.Parse(line));
        }

        private static Intents IntentsFrom(object value)
        {
            if (value == null) return Intents.None;
            if (!(value is Dictionary<string, object> obj))
            {
                throw new FormatException("each input must be a JSON object");
            }
            return new Intents(Json.Flag(obj, "nudge_up"), Json.Flag(obj, "nudge_down"));
        }

        /// `{"version": 1, "inputs": [ {..}, .. ]}`. A missing path, or `-`,
        /// means "nothing held on every tick".
        private static List<Intents> ReadScript(string path)
        {
            var inputs = new List<Intents>();
            if (string.IsNullOrEmpty(path) ||
                string.Equals(path, "-", StringComparison.Ordinal))
            {
                return inputs;
            }
            if (!File.Exists(path))
            {
                throw new FileNotFoundException($"input script not found: {path}", path);
            }

            if (!(Json.Parse(File.ReadAllText(path)) is Dictionary<string, object> root))
            {
                throw new FormatException("an input script must be a JSON object");
            }
            if (!root.TryGetValue("version", out var version) ||
                !(version is double number) || (int)number != 1)
            {
                throw new FormatException("an input script must declare \"version\": 1");
            }
            if (root.TryGetValue("inputs", out var raw))
            {
                if (!(raw is List<object> list))
                {
                    throw new FormatException("\"inputs\" must be an array");
                }
                foreach (var element in list) inputs.Add(IntentsFrom(element));
            }
            return inputs;
        }

        // ------------------------------------------------------------------
        // Command line
        // ------------------------------------------------------------------

        private static string OptionalArg(string name)
        {
            var args = Environment.GetCommandLineArgs();
            for (int i = 0; i < args.Length - 1; i++)
            {
                if (string.Equals(args[i], name, StringComparison.Ordinal)) return args[i + 1];
            }
            return null;
        }

        private static string StringArg(string name) =>
            OptionalArg(name)
            ?? throw new ArgumentException($"missing required argument {name}", nameof(name));

        private static ulong ULongArg(string name) =>
            ulong.Parse(StringArg(name), CultureInfo.InvariantCulture);

        private static int IntArg(string name) =>
            int.Parse(StringArg(name), CultureInfo.InvariantCulture);
    }
}

File: eval/starters/rust/crates/sim/src/bin/probe.rs
//! Headless observation of a run, in two forms.
//!
//! * `probe stream SEED` — a long-lived process. Writes one trace line for the
//!   world before any tick has run, then reads one JSON input object per line
//!   from stdin, steps exactly one tick per line, and writes one trace line per
//!   tick. Flushes after every line, so a driver can read the state it just
//!   caused and decide what to send next. EOF or a line reading `quit` exits 0.
//! * `probe file SEED TICKS SCRIPT OUT` — the batch form. Replays a whole
//!   script and writes the trace to a file.
//!
//! **stdout carries nothing but trace lines.** Everything else goes to stderr:
//! a stray banner on stdout desynchronises whatever is reading it.
//!
//! Both forms are the same loop over [`sim::replay::headless_app`], so a trace
//! is exactly what the determinism tests hash — no second implementation of the
//! tick loop to drift out of sync.

use std::io::{self, BufRead, Write};
use std::process::ExitCode;

use bevy_app::App;
use bevy_ecs::prelude::With;
use sim::replay::headless_app;
use sim::{Intents, Marker, Position, Tick, TickEvents, Velocity, script, state_hash};

fn main() -> ExitCode {
    match run() {
        Ok(()) => ExitCode::SUCCESS,
        Err(why) => {
            eprintln!("probe: {why}");
            ExitCode::FAILURE
        }
    }
}

const USAGE: &str = "usage:\n  \
     probe stream SEED\n  \
     probe file SEED TICKS SCRIPT OUT   (SCRIPT may be '-' for no input)";

fn run() -> Result<(), String> {
    let args: Vec<String> = std::env::args().skip(1).collect();
    match args.iter().map(String::as_str).collect::<Vec<_>>()[..] {
        ["stream", seed] => stream(parse_u64(seed, "SEED")?),
        ["file", seed, ticks, path, out] => batch(
            parse_u64(seed, "SEED")?,
            usize::try_from(parse_u64(ticks, "TICKS")?).map_err(|_| "TICKS is too large")?,
            path,
            out,
        ),
        _ => Err(format!("bad arguments.\n{USAGE}")),
    }
}

fn parse_u64(text: &str, what: &str) -> Result<u64, String> {
    text.parse()
        .map_err(|_| format!("{what} must be a non-negative integer, got {text:?}"))
}

/// Interactive form: one input line in, one trace line out, flushed every time.
fn stream(seed: u64) -> Result<(), String> {
    let mut app = headless_app(seed);
    let stdout = io::stdout();
    let mut out = stdout.lock();

    emit(&mut out, &mut app)?;

    for line in io::stdin().lock().lines() {
        let line = line.map_err(|e| format!("could not read stdin: {e}"))?;
        if line.trim() == "quit" {
            break;
        }
        let intents = script::parse_intents(&line)?;
        step(&mut app, intents);
        emit(&mut out, &mut app)?;
    }
    Ok(())
}

/// Batch form: replay a whole script, then write the trace in one go.
fn batch(seed: u64, ticks: usize, script_path: &str, out_path: &str) -> Result<(), String> {
    let inputs = load_script(script_path)?;
    let mut app = headless_app(seed);
    let mut trace = String::new();

    for tick in 0..ticks {
        step(&mut app, inputs.get(tick).copied().unwrap_or_default());
        trace.push_str(&trace_line(&mut app)?);
        trace.push('\n');
    }

    let reached = app.world().resource::<Tick>().0;
    if reached != ticks as u64 {
        return Err(format!(
            "asked for {ticks} ticks, the simulation reached {reached}"
        ));
    }

    std::fs::write(out_path, trace).map_err(|e| format!("could not write {out_path}: {e}"))
}

fn load_script(path: &str) -> Result<Vec<Intents>, String> {
    if path == "-" || path.is_empty() {
        return Ok(Vec::new());
    }
    let text = std::fs::read_to_string(path).map_err(|e| format!("could not read {path}: {e}"))?;
    script::parse_script(&text).map_err(|why| format!("{path}: {why}"))
}

fn step(app: &mut App, intents: Intents) {
    *app.world_mut().resource_mut::<Intents>() = intents;
    app.update();
}

fn emit(out: &mut impl Write, app: &mut App) -> Result<(), String> {
    let line = trace_line(app)?;
    writeln!(out, "{line}").map_err(|e| format!("could not write to stdout: {e}"))?;
    out.flush()
        .map_err(|e| format!("could not flush stdout: {e}"))
}

/// One JSON Lines record: `{"tick": N, "hash": "0x…", "state": {…}, "events": […]}`.
///
/// `state` is game-defined. Expose the values that describe what the game is
/// doing right now, keep the shape stable, and keep every number finite.
fn trace_line(app: &mut App) -> Result<String, String> {
    let tick = app.world().resource::<Tick>().0;
    let events = app.world().resource::<TickEvents>().events.clone();
    let hash = state_hash(app.world_mut());

    let mut markers = app
        .world_mut()
        .query_filtered::<(&Position, &Velocity), With<Marker>>();
    let (position, velocity) = markers
        .iter(app.world())
        .next()
        .ok_or("the world contains no marker to report on")?;

    let state = format!(
        r#"{{"marker": {{"x": {}, "y": {}, "vx": {}, "vy": {}}}}}"#,
        number(position.0.x)?,
        number(position.0.y)?,
        number(velocity.0.x)?,
        number(velocity.0.y)?,
    );

    let events = events
        .iter()
        .map(|name| json_string(name))
        .collect::<Vec<_>>()
        .join(", ");

    Ok(format!(
        r#"{{"tick": {tick}, "hash": "0x{hash:016x}", "state": {state}, "events": [{events}]}}"#
    ))
}

/// f32 as a finite JSON number. `{:?}` on an f32 is the shortest decimal that
/// round-trips and always carries a `.` or an exponent, so it is both exact and
/// valid JSON. NaN and infinity are errors, not values: they mean the
/// simulation has already diverged and a consumer cannot parse them anyway.
fn number(value: f32) -> Result<String, String> {
    if value.is_finite() {
        Ok(format!("{value:?}"))
    } else {
        Err(format!(
            "the simulation produced a non-finite value ({value}); a trace must \
             contain only finite JSON numbers"
        ))
    }
}

fn json_string(text: &str) -> String {
    let mut out = String::with_capacity(text.len() + 2);
    out.push('"');
    for ch in text.chars() {
        match ch {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            c if (c as u32) < 0x20 => out.push_str(&format!("\\u{:04x}", c as u32)),
            c => out.push(c),
        }
    }
    out.push('"');
    out
}
